docs/spec/overload.rst

```diff
@@ -99,8 +109,349 @@ variable is not sufficient.
+one is present, an error should be reported.
+
+The ``@overload``-decorated definitions must be followed by an overload
+implementation, which does not include an ``@overload`` decorator. Type
+checkers should report an error or warning if an implementation is missing.
+Overload definitions within stub files, protocols, and on abstract methods
+within abstract base classes are exempt from this check.
+
+If one overload signature is decorated with ``@staticmethod`` or
+``@classmethod``, all overload signatures must be similarly decorated. The
+implementation, if present, must also have a consistent decorator. Type
+checkers should report an error if these conditions are not met.
+
+If a ``@final`` or ``@override`` decorator is supplied for a function with
+overloads, the decorator should be applied only to the overload implementation
+if it is present. If an overload implementation isn't present (for example, in
+a stub file), the ``@final`` or ``@override`` decorator should be applied only
+to the first overload. Type checkers should enforce these rules and generate
+an error when they are violated. If a ``@final`` or ``@override`` decorator
+follows these rules, a type checker should treat the decorator as if it is
+present on all overloads.
+
+Overloads are allowed to use a mixture of ``async def`` and ``def`` statements
+within the same overload definition. Type checkers should convert
+``async def`` statements to a non-async signature (wrapping the return
+type in a ``Coroutine``) before testing for implementation consistency
+and overlapping overloads (described below).
+
+
+Implementation consistency
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If an overload implementation is defined, type checkers should validate
+that it is consistent with all of its associated overload signatures.
+The implementation should accept all potential sets of arguments
+that are accepted by the overloads and should produce all potential return
+types produced by the overloads. In typing terms, this means the input
+signature of the implementation should be :term:<assignable> to the input
+signatures of all overloads, and the return type of all overloads should be
+assignable to the return type of the implementation.
+
+If the implementation is inconsistent with its overloads, a type checker
+should report an error::
+
+  @overload
+  def func(x: str, /) -> str: ...
+  @overload
+  def func(x: int) -> int: ...
+
+  # This implementation is inconsistent with the second overload
+  # because it does not accept a keyword argument ``x`` and the
+  # the overload's return type ``int`` is not assignable to the
+  implementation's return type ``str``.
+  def func(x: int | str, /) -> str:
+    return str(x)
+
+When a type checker checks the implementation for consistency with overloads,
+it should first apply any transforms that change the effective type of the
+implementation including the presence of a ``yield`` statement in the
```

Comment:
Is there a circumstance in which the presence of a `yield` in the body of a function changes the meaning of its return type annotation? My understanding is that it does not: a generator must be manually annotated with a `Generator` or `Iterator` return type. So I'm not sure how to write a test for this mention of "presence of a `yield` statement." Should this be removed from the text?

Reply:
You're correct, the presence of a `yield` statement doesn't change the effective type of the return type annotation. It does affect the inferred return type if the type checker (like pyright) implements return type inference. I think I meant to say "`async` keyword" rather than "`yield` statement". This should be changed in the spec.

Reply:
`yield` does change the meaning of the return type for coroutines, right?

Reply:
The behavior of `yield` and its effects on the return type are not currently specified in the typing spec, but all major type checkers do change the meaning of the return type in this case.

Reply:
> `yield` does change the meaning of the return type for coroutines, right?

Hmm, yes, I think it does. It requires you to annotate with `typing.AsyncGenerator`, which is _not_ implicitly wrapped in `Coroutine`, like the return type of an `async def` would be if it doesn't yield. So presence of `yield` in the body of an `async def` does have to be considered by a type checker when interpreting the return type. Which I guess means no change to the text here, and maybe a test should be added for this case.